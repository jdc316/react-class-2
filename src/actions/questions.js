import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { updateUser } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const UPDATE_QUESTION = "UPDATE_QUESTION";

export function receiveQuestions(questions) { 
    return {
        type: RECEIVE_QUESTIONS,
        questions
    };
}

export function addQuestion(question) {
    return {
        type: ADD_QUESTION,
        question,
    };
}

export function updateQuestion(question) {
    return {
        type: UPDATE_QUESTION,
        question,
    };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
    return (dispatch, getState) => {
        const { authedUser } = getState();

        dispatch(showLoading());

        return saveQuestion({
            author: authedUser.id,
            optionOneText,
            optionTwoText
        })
            .then((question) => dispatch(addQuestion(question)))
            .then(() => dispatch(hideLoading())
        );
    }
}

export function handleAnswerQuestion(qid, answer) {
    return (dispatch, getState) => {
        const { authedUser, users, questions } = getState();

        dispatch(showLoading());
        let user = users[authedUser.id];

        user = {
            ...user,
            answers: {
                ...user.answers,
                [qid]: answer
            }
        }

        let question = questions[qid];

        question = {
            ...question,
            [answer]: {
                ...question[answer],
                votes: question[answer].votes.concat(authedUser.id)
            }
        }

        return saveQuestionAnswer({
            authedUser: authedUser.id,
            qid,
            answer
        })
        .then(() => {
            dispatch(updateUser(user));
        })
        .then(() => dispatch(updateQuestion(question)))
        .then(() => dispatch(hideLoading()));
        
    }
}