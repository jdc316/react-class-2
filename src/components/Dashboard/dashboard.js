import { connect } from "react-redux";
import QuestionsList from "../QuestionList/questionsList";

const Dashboard = () => {
    return (
        <div>
            <h2 className="center">Home</h2>
            <QuestionsList />
        </div>
    );
}

export default connect()(Dashboard);