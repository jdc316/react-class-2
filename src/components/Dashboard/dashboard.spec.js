import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from "../../reducers";
import middleware from "../../middleware";
import { MemoryRouter } from 'react-router-dom';
import { screen } from '@testing-library/react';
import Dashboard from './dashboard';
import { setAuthedUser } from '../../actions/authedUser';
import { handleInitialData } from '../../actions/shared';
import * as router from 'react-router'

describe('Dashboard', () => {
  const store = createStore(reducer, middleware);
  const navigate = jest.fn();

  beforeEach(() => {
    jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate)
  });
  
  test('renders Dashboard', async () => {
    await store.dispatch(handleInitialData());
    await store.dispatch(setAuthedUser({id: "sarahedo"}));
        
    render( 
      <MemoryRouter>
        <Provider store={store}>
          <Dashboard />
        </Provider>
      </MemoryRouter> 
    );
  
    expect(screen.getByText(/Home/i)).toBeInTheDocument();
  });

  test('matches snapshot', async () => {
    await store.dispatch(handleInitialData());
    await store.dispatch(setAuthedUser({id: "sarahedo"}));

    // eslint-disable-next-line testing-library/render-result-naming-convention
    var app = render( 
      <MemoryRouter>
        <Provider store={store}>
          <Dashboard />
        </Provider>
      </MemoryRouter> 
    );
    
    expect(app).toMatchSnapshot();
  });

  test('can open Question', async () => {
    await store.dispatch(handleInitialData());
    await store.dispatch(setAuthedUser({id: "sarahedo"}));
        
    render( 
      <MemoryRouter>
        <Provider store={store}>
          <Dashboard />
        </Provider>
      </MemoryRouter> 
    );

    let button = screen.getAllByText('Show')[0];
    fireEvent.click(button);

    expect(navigate).toHaveBeenCalled();
  })
})
  