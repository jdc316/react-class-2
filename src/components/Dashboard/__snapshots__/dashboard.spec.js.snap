// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2
          class="center"
        >
          Home
        </h2>
        <div>
          <div>
            <h3
              class="center"
            >
              New Questions
            </h3>
            <div
              class="question-container"
            >
              <div
                class="question"
              >
                <span
                  class="center"
                >
                  <b>
                    mtsamis
                  </b>
                </span>
                <span
                  class="center"
                >
                  4/30/2017, 3:16:07 PM
                </span>
                <button
                  class="btn-md green"
                >
                  Show
                </button>
              </div>
              <div
                class="question"
              >
                <span
                  class="center"
                >
                  <b>
                    tylermcginnis
                  </b>
                </span>
                <span
                  class="center"
                >
                  3/15/2017, 8:09:27 AM
                </span>
                <button
                  class="btn-md green"
                >
                  Show
                </button>
              </div>
            </div>
          </div>
          <div>
            <h3
              class="center"
            >
              Done
            </h3>
            <div
              class="question-container"
            >
              <div
                class="question"
              >
                <span
                  class="center"
                >
                  <b>
                    sarahedo
                  </b>
                </span>
                <span
                  class="center"
                >
                  3/3/2017, 5:22:47 PM
                </span>
                <button
                  class="btn-md green"
                >
                  Show
                </button>
              </div>
              <div
                class="question"
              >
                <span
                  class="center"
                >
                  <b>
                    tylermcginnis
                  </b>
                </span>
                <span
                  class="center"
                >
                  12/24/2016, 6:42:47 AM
                </span>
                <button
                  class="btn-md green"
                >
                  Show
                </button>
              </div>
              <div
                class="question"
              >
                <span
                  class="center"
                >
                  <b>
                    mtsamis
                  </b>
                </span>
                <span
                  class="center"
                >
                  7/14/2016, 3:02:47 AM
                </span>
                <button
                  class="btn-md green"
                >
                  Show
                </button>
              </div>
              <div
                class="question"
              >
                <span
                  class="center"
                >
                  <b>
                    sarahedo
                  </b>
                </span>
                <span
                  class="center"
                >
                  6/28/2016, 10:21:12 PM
                </span>
                <button
                  class="btn-md green"
                >
                  Show
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2
        class="center"
      >
        Home
      </h2>
      <div>
        <div>
          <h3
            class="center"
          >
            New Questions
          </h3>
          <div
            class="question-container"
          >
            <div
              class="question"
            >
              <span
                class="center"
              >
                <b>
                  mtsamis
                </b>
              </span>
              <span
                class="center"
              >
                4/30/2017, 3:16:07 PM
              </span>
              <button
                class="btn-md green"
              >
                Show
              </button>
            </div>
            <div
              class="question"
            >
              <span
                class="center"
              >
                <b>
                  tylermcginnis
                </b>
              </span>
              <span
                class="center"
              >
                3/15/2017, 8:09:27 AM
              </span>
              <button
                class="btn-md green"
              >
                Show
              </button>
            </div>
          </div>
        </div>
        <div>
          <h3
            class="center"
          >
            Done
          </h3>
          <div
            class="question-container"
          >
            <div
              class="question"
            >
              <span
                class="center"
              >
                <b>
                  sarahedo
                </b>
              </span>
              <span
                class="center"
              >
                3/3/2017, 5:22:47 PM
              </span>
              <button
                class="btn-md green"
              >
                Show
              </button>
            </div>
            <div
              class="question"
            >
              <span
                class="center"
              >
                <b>
                  tylermcginnis
                </b>
              </span>
              <span
                class="center"
              >
                12/24/2016, 6:42:47 AM
              </span>
              <button
                class="btn-md green"
              >
                Show
              </button>
            </div>
            <div
              class="question"
            >
              <span
                class="center"
              >
                <b>
                  mtsamis
                </b>
              </span>
              <span
                class="center"
              >
                7/14/2016, 3:02:47 AM
              </span>
              <button
                class="btn-md green"
              >
                Show
              </button>
            </div>
            <div
              class="question"
            >
              <span
                class="center"
              >
                <b>
                  sarahedo
                </b>
              </span>
              <span
                class="center"
              >
                6/28/2016, 10:21:12 PM
              </span>
              <button
                class="btn-md green"
              >
                Show
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
